doc/tutorial/tutorial.md

```diff
@@ -1,40 +1,650 @@
+// Rule
+Phrases(w1 ++ " " ++ w2) :- Word1(w1, cat), Word2(w2, cat).
+```
+
+This may not look like a typical "Hello, world!" program, yet that is exactly what it is.  Let's
+dissect this code snippet.  First, note the use of C++ style comments (`//` and `/* */`).
+
+Comments aside, this program contains several DDlog declarations: a data type declaration, three
+relations, and a rule.  We discuss the type system in more detail below.  For the moment, think of
+the `Category` type as a C-style enum with two possible values: `CategoryStarWars` and
+`CategoryOther`.
+
+Datalog relations are similar to database tables in that they store sets of records.  `Word1` and
+`Word2` are two *input* relations in our example.  Input relations (aka *extensional* relations in
+the Datalog world) are populated by user-provided ground facts.  The other kind of relations are
+*derived* (or *intensional*) relations that are computed based on rules.  The two kinds do not
+overlap: a relation must be populated exclusively by either ground or derived facts.  The `input`
+qualifier is used to declare an input relation; relations declared without this qualifier are
+derived relations.
+
+`Word1` and `Word2` relations have the same record type, including a `word` field of type `string`
+and a `cat` field of type `Category`.  We would like to combine strings from `Word1` and `Word2`
+relations into phrases and store them in the `Phrases` relation.  However, we only want to combine
+words that belong to the same category.  This is captured by the DDlog rule in the end of the
+listing.  The rule can be read as follows: *for every `w1`, `w2`, and `cat`, such that `(w1, cat)`
+is in `Word1` and `(w2, cat)` is in `Word2`, there is a record `w1 ++ " " ++ w2` in Phrases*, where,
+`++` is a string concatenation operator.
+
+Clauses of the form `<Relation_name>(<arguments>)` are called *atoms*.  An atom is true iff the
+relation contains the specified record.  The atom to the left of `:-` is the *head* of the rule;
+atoms to the right of `:-` form the *body* of the rule.  Commas in the body of the rule represent
+logical conjunctions, i.e., if all atoms in the body are true, then the head of the rule holds.
+Finally, note the mandatory dot in the end of the rule.
+
+The above rule computes an inner join of `Word1` and `Word2` relations on the `cat` field.
+
+### Capitalization rules
+
+1. A relation name must start with an upper-case letter.
+
+1. A variable name must start with a lower-case letter or underscore
+(`_`).
+
+1. A type name can start with an upper-case or a lower-case letter or
+an underscore.
+
+### Ordering of declarations
+
+DDlog does not care about the order in which various program entities (types, relations, rules,
+functions) are declared.  Reordering declarations in the above program does not have any effect on
+program semantics.
+
+## Compiling the example
+
+Run `stack test --ta '-p playpen'` to compile the "Hello, world!" program.  This command performs
```

Comment:
I think that either an ASCII diagram or a png file would be more compact as an explanation.